### ISSUE TITLE
Need clarity on Tapes and gradient tracking

### ISSUE THREAD

**Issue by vikigenius** [coreylowman/dfdx#112]
I don't know if it's just me, but I found it hard to wrap my head around the usage of Tapes for gradient tracking.

It might be a good idea to describe it in terms of analogous concepts for those coming from other frameworks in Python like Pytorch/Tensorflow.

For instance, is NoTape the equivalent of setting `requires_grad=False` in Pytorch?

Pytorch has the concept of parameters vs buffers. A buffer is basically a Tensor that is known to not track gradients but still need to be registered for moving to devices, saving to disk, changing precision etc..

In our case is it sufficient for us to just not update the tensors in the `CanUpdateWithGradients` implementation.

For example consider a partial implementation of BatchNorm

```rust
use crate::prelude::*;


pub struct BatchNorm1D<const M: usize> {
    pub gamma: Tensor1D<M, NoneTape>,
    pub beta: Tensor1D<M, NoneTape>,
    pub epsilon: f32,
    pub running_mean: Tensor1D<M, NoneTape>,
    pub running_var: Tensor1D<M, NoneTape>,
    pub num_batches_tracked: Tensor0D<NoneTape>,
}


impl<const M: usize> CanUpdateWithGradients for BatchNorm1D<M> {
    /// Updates [Self::gamma] and [Self::beta].
    fn update<G: GradientProvider>(&mut self, grads: &mut G) {
        self.gamma.update(grads);
        self.beta.update(grads);
    }
}

```

Is this enough for gradients not to be tracked for `running_mean`, `running_var` and `num_batches_tracked` ?

**Comment by vikigenius** [comment]
I feel like there should be a way to prevent gradients being tracked in `forward` when I update the running mean, but I am not sure how that can be achieved

**Comment by coreylowman** [comment]
Good question - the usage of NoneTape for tensors just means that those tensors will never *own* the tape. However they still can be tracked.

Where this comes into play is indeed when you implement the Module trait (which contains forward). You can specify the tape that the input tensor has as a generic one, which will handle both OwnsTape & NoneTape.

```rust
impl<TAPE: Tape, ...> Module<Tensor1D<..., TAPE>> for BatchNorm1D<...> {
     type Output = Tensor1D<..., TAPE>; // NOTE: the output contains the same TAPE as the input does
     fn forward(&self, input: Tensor1D<..., TAPE>) -> Self::Output {
          ...
     }
}
```

If you want you could even have separate impls for OwnsTape and NoneTape, which could be useful for BatchNorm
```rust
/// this impl is for when there is no tape in the input - no gradients are being tracked here!
impl<...> Module<Tensor1D<..., NoneTape>> for BatchNorm1D<...> {
     type Output = Tensor1D<..., NoneTape>;
     fn forward(&self, input: Tensor1D<..., NoneTape>) -> Self::Output {
          ...
     }
}

/// this impl is for when there is a tape in the input - gradients ARE being tracked here!
impl<...> Module<Tensor1D<..., OwnsTape>> for BatchNorm1D<...> {
     type Output = Tensor1D<..., OwnsTape>;
     fn forward(&self, input: Tensor1D<..., OwnsTape>) -> Self::Output {
          ...
     }
}
```

Does that help?

**Comment by vikigenius** [comment]
Thanks that does help. Closing this issue now. I will try to contribute to the docs after I figure out batch norm.
